# Exercise

Download heart disease dataset heart.csv and do following, (credits of dataset: https://www.kaggle.com/fedesoriano/heart-failure-prediction)

1. Load heart disease dataset in pandas dataframe
2. Remove outliers using Z score. Usual guideline is to remove anything that has Z score > 3 formula or Z score < -3
3. Convert text columns to numbers using label encoding and one hot encoding
4. Apply scaling
5. Build a classification model using support vector machine. Use standalone model as well as Bagging model and check if you see any difference in the performance.
6. Now use decision tree classifier. Use standalone model as well as Bagging and check if you notice any difference in performance
7. Comparing performance of svm and decision tree classifier figure out where it makes most sense to use bagging and why. Use internet to figure out in what conditions bagging works the best.

In [1]:
import pandas as pd

df = pd.read_csv('20_heart.csv')
df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  ...  MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease
0   40   M           ATA        140          289  ...    172              N      0.0       Up             0
1   49   F           NAP        160          180  ...    156              N      1.0     Flat             1
2   37   M           ATA        130          283  ...     98              N      0.0       Up             0
3   48   F           ASY        138          214  ...    108              Y      1.5     Flat             1
4   54   M           NAP        150          195  ...    122              N      0.0       Up             0

[5 rows x 12 columns]

In [2]:
df.shape

(918, 12)

In [3]:
df.describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR     Oldpeak  HeartDisease
count  918.000000  918.000000   918.000000  918.000000  918.000000  918.000000    918.000000
mean    53.510893  132.396514   198.799564    0.233115  136.809368    0.887364      0.553377
std      9.432617   18.514154   109.384145    0.423046   25.460334    1.066570      0.497414
min     28.000000    0.000000     0.000000    0.000000   60.000000   -2.600000      0.000000
25%     47.000000  120.000000   173.250000    0.000000  120.000000    0.000000      0.000000
50%     54.000000  130.000000   223.000000    0.000000  138.000000    0.600000      1.000000
75%     60.000000  140.000000   267.000000    0.000000  156.000000    1.500000      1.000000
max     77.000000  200.000000   603.000000    1.000000  202.000000    6.200000      1.000000

In [4]:
# Treat Outliers

In [7]:
df[df.Cholesterol>(df.Cholesterol.mean()+3*df.Cholesterol.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  ...  MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease
76    32   M           ASY        118          529  ...    130              N      0.0     Flat             1
149   54   M           ASY        130          603  ...    125              Y      1.0     Flat             1
616   67   F           NAP        115          564  ...    160              N      1.6     Flat             0

[3 rows x 12 columns]

In [8]:
df.shape

(918, 12)

In [9]:
df1 = df[df.Cholesterol<=(df.Cholesterol.mean()+3*df.Cholesterol.std())]
df1.shape

(915, 12)

In [10]:
df[df.MaxHR>(df.MaxHR.mean()+3*df.MaxHR.std())]

Empty DataFrame
Columns: [Age, Sex, ChestPainType, RestingBP, Cholesterol, FastingBS, RestingECG, MaxHR, ExerciseAngina, Oldpeak, ST_Slope, HeartDisease]
Index: []

In [11]:
df[df.FastingBS>(df.FastingBS.mean()+3*df.FastingBS.std())]

Empty DataFrame
Columns: [Age, Sex, ChestPainType, RestingBP, Cholesterol, FastingBS, RestingECG, MaxHR, ExerciseAngina, Oldpeak, ST_Slope, HeartDisease]
Index: []

In [12]:
df[df.Oldpeak>(df.Oldpeak.mean()+3*df.Oldpeak.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  ...  MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease
166   50   M           ASY        140          231  ...    140              Y      5.0     Flat             1
702   59   M            TA        178          270  ...    145              N      4.2     Down             0
771   55   M           ASY        140          217  ...    111              Y      5.6     Down             1
791   51   M           ASY        140          298  ...    122              Y      4.2     Flat             1
850   62   F           ASY        160          164  ...    145              N      6.2     Down             1
900   58   M           ASY        114          318  ...    140              N      4.4     Down             1

[6 rows x 12 columns]

In [13]:
df2 = df1[df1.Oldpeak<=(df1.Oldpeak.mean()+3*df1.Oldpeak.std())]
df2.shape

(909, 12)

In [14]:
df[df.RestingBP>(df.RestingBP.mean()+3*df.RestingBP.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  ...  MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease
109   39   M           ATA        190          241  ...    106              N      0.0       Up             0
241   54   M           ASY        200          198  ...    142              Y      2.0     Flat             1
365   64   F           ASY        200            0  ...    140              Y      1.0     Flat             1
399   61   M           NAP        200            0  ...     70              N      0.0     Flat             1
592   61   M           ASY        190          287  ...    150              Y      2.0     Down             1
732   56   F           ASY        200          288  ...    133              Y      4.0     Down             1
759   54   M           ATA        192          283  ...    195              N      0.0       Up             1

[7 rows x 12 columns]

In [16]:
df3 = df2[df2.RestingBP<=(df2.RestingBP.mean()+3*df2.RestingBP.std())]
df3.shape

(902, 12)

In [17]:
df.ChestPainType.unique()

array(['ATA', 'NAP', 'ASY', 'TA'], dtype=object)

In [18]:
df.RestingECG.unique()

array(['Normal', 'ST', 'LVH'], dtype=object)

In [19]:
df.ExerciseAngina.unique()

array(['N', 'Y'], dtype=object)

In [20]:
df.ST_Slope.unique()

array(['Up', 'Flat', 'Down'], dtype=object)

Handle text columns using label encoding and one hot encoding

In [23]:
df4 = df3.copy()

df4.ExerciseAngina.replace({
    'N':0,
    'Y':1
}, inplace = True)

df4.ST_Slope.replace({
    'Up':1,
    'Flat':2,
    'Down':3
}, inplace=True)

df4.RestingECG.replace({
    'Normal':1,
    'ST':2,
    'LVH':3
}, inplace=True)

df4.head()

C:\Users\hassa\AppData\Local\Temp\ipykernel_6928\3723989606.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df4.ExerciseAngina.replace({
C:\Users\hassa\AppData\Local\Temp\ipykernel_6928\3723989606.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df4.ExerciseAngina.replace({
C:\Users\hassa\App

Age Sex ChestPainType  RestingBP  Cholesterol  ...  MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease
0   40   M           ATA        140          289  ...    172               0      0.0         1             0
1   49   F           NAP        160          180  ...    156               0      1.0         2             1
2   37   M           ATA        130          283  ...     98               0      0.0         1             0
3   48   F           ASY        138          214  ...    108               1      1.5         2             1
4   54   M           NAP        150          195  ...    122               0      0.0         1             0

[5 rows x 12 columns]

In [24]:
df5 = pd.get_dummies(df4,drop_first=True)
df5.head()

Age  RestingBP  Cholesterol  FastingBS  ...  Sex_M  ChestPainType_ATA  ChestPainType_NAP  ChestPainType_TA
0   40        140          289          0  ...   True               True              False             False
1   49        160          180          0  ...  False              False               True             False
2   37        130          283          0  ...   True               True              False             False
3   48        138          214          0  ...  False              False              False             False
4   54        150          195          0  ...   True              False               True             False

[5 rows x 14 columns]

In [25]:
x = df5.drop('HeartDisease',axis='columns')
y = df5.HeartDisease

x.head()

Age  RestingBP  Cholesterol  FastingBS  ...  Sex_M  ChestPainType_ATA  ChestPainType_NAP  ChestPainType_TA
0   40        140          289          0  ...   True               True              False             False
1   49        160          180          0  ...  False              False               True             False
2   37        130          283          0  ...   True               True              False             False
3   48        138          214          0  ...  False              False              False             False
4   54        150          195          0  ...   True              False               True             False

[5 rows x 13 columns]

In [26]:
from sklearn.preprocessing import StandardScaler

scalar = StandardScaler()
x_scaled = scalar.fit_transform(x)
x_scaled

array([[-1.42896269,  0.46089071,  0.85238015, ...,  2.06757196,
        -0.53547478, -0.22914788],
       [-0.47545956,  1.5925728 , -0.16132855, ..., -0.4836591 ,
         1.86750159, -0.22914788],
       [-1.74679706, -0.10495034,  0.79657967, ...,  2.06757196,
        -0.53547478, -0.22914788],
       ...,
       [ 0.37209878, -0.10495034, -0.61703246, ..., -0.4836591 ,
        -0.53547478, -0.22914788],
       [ 0.37209878, -0.10495034,  0.35947592, ...,  2.06757196,
        -0.53547478, -0.22914788],
       [-1.64085227,  0.3477225 , -0.20782894, ..., -0.4836591 ,
         1.86750159, -0.22914788]])

In [29]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(x_scaled, y, test_size=0.2, random_state=20)

In [31]:
X_train.shape

(721, 13)

In [32]:
X_test.shape

(181, 13)

Train a model using standalone support vector machine and then using bagging

In [33]:
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score

scores = cross_val_score(SVC(), x, y, cv=5)
scores.mean()

np.float64(0.6906445672191528)

Use bagging now with svm

In [36]:
from sklearn.ensemble import BaggingClassifier

bag_model = BaggingClassifier(estimator=SVC(), n_estimators=100, max_samples=0.8, random_state=0)
scores = cross_val_score(bag_model, x, y, cv=5)
scores.mean()

np.float64(0.6839656230816453)

As you can see above, using bagging in case of SVM doesn't make much difference in terms of model accuracy. Bagging is effective when we have high variance and instable model such as decision tree. Let's explore how bagging changes the performance for a decision tree classifier.

Train a model using decision tree and then using bagging

In [37]:
from sklearn.tree import DecisionTreeClassifier

scores = cross_val_score(DecisionTreeClassifier(random_state=0), x, y, cv=5)
scores.mean()

np.float64(0.7182872928176797)

Use bagging now with decision tree

In [38]:
bag_model = BaggingClassifier(
    estimator=DecisionTreeClassifier(random_state=0),
    n_estimators=100,
    max_samples=0.9,
    oob_score=True,
    random_state=0
)

scores = cross_val_score(bag_model, x, y, cv=5)
scores.mean()

np.float64(0.8036893799877225)

You can see that with bagging the score improved from 71.93% to 80.37%

# Train a model using Random Forest which itself uses bagging underneath

In [39]:
from sklearn.ensemble import RandomForestClassifier

scores = cross_val_score(RandomForestClassifier(), x, y, cv=5)
scores.mean()

np.float64(0.823670963781461)

Random forest gave even a better performance with 82.3% as score. Underneath it used bagging where it sampled not only data rows but also the columns (or features)